modin/core/execution/ray/common/deferred_execution.py

```diff
@@ -0,0 +1,708 @@
+    single batch (i.e. using a single remote call) and the results are saved in all
+    the nodes with ref_counter > 1.
+
+    Parameters
+    ----------
+    data : ObjectRefType or DeferredExecution
+        The execution input.
+    func : callable or ObjectRefType
+        A function to be executed.
+    args : iterable
+        Additional positional arguments to be passed in `func`.
+    kwargs : dict
+        Additional keyword arguments to be passed in `func`.
+    num_returns : int
+        The number of the return values.
+    flat_args : bool, optional
+        True means that there are no lists or DeferredExecution objects in `args`.
+        In this case, no arguments processing is performed and `args` is passed
+        to the remote method as is.
+    flat_kwargs : bool, optional
+        The same as `flat_args` but for the `kwargs` values.
+    """
+
+    def __init__(
+        self,
+        data: Union[
+            ObjectRefType,
+            "DeferredExecution",
+            List[Union[ObjectRefType, "DeferredExecution"]],
+        ],
+        func: Union[Callable, ObjectRefType],
+        args: Iterable[Any],
+        kwargs: Dict[str, Any],
+        num_returns=1,
+        flat_args: Optional[bool] = None,
+        flat_kwargs: Optional[bool] = None,
+    ):
+        ref = DeferredExecution._ref
+        if flat_args is None:
+            if has_list_or_de(args):
+                flat_args = False
+                for a in args:
+                    ref(a)
+            else:
+                flat_args = True
+        if flat_kwargs is None:
+            if has_list_or_de(kwargs.values()):
+                flat_kwargs = False
+                for a in kwargs.values():
+                    ref(a)
+            else:
+                flat_kwargs = True
+        ref(data)
+        self.data = data
+        self.task = (func, args, kwargs, flat_args, flat_kwargs)
+        self.num_returns = num_returns
+        self.ref_counter = 0
+
+    @staticmethod
+    def _ref(obj):
+        """
+        Increment the `ref_counter` if `obj` is a `DeferredExecution`.
+
+        Parameters
+        ----------
+        obj : Any
+        """
+        if isinstance(obj, DeferredExecution):
+            obj.ref_count(1)
+        elif isinstance(obj, ListOrTuple):
+            ref = DeferredExecution._ref
+            for o in obj:
+                ref(o)
+
+    def exec(
+        self,
+    ) -> Tuple[ObjectRefOrListType, Union["MetaList", List], Union[int, List[int]]]:
+        """
+        Execute this task, if required.
+
+        Returns
+        -------
+        tuple
+            The execution result, MetaList, containing the length, width and
+            the worker's ip address (the last value in the list) and the values
+            offset in the list. I.e. length = meta_list[offset],
+            width = meta_list[offset + 1], ip = meta_list[-1].
+        """
+        if not self.has_result:
```

Comment:
let's invert this condition so the main code would have fewer indents
```suggestion
        if self.has_result:
            return self.data, self.meta, self.meta_off
        ...
```